### ISSUE TITLE
/plotme auto bugs in 0.17

### ISSUE THREAD

**Issue by divad** [WorldCretornica/PlotMe-Core#322]
Version: 0.17.4-beta

There are several problems with /plotme auto

When using /plotme auto in a plot world the plotme plugin tells the user "No plot found" but then, several seconds later (i.e its very slow), it then teleports the player to their new plot. It also seems to ignore the [world] parameter and always gives you a plot in the plot world you're already in.

If you're not in a plot world and use /plotme auto <world> then it says "No plot found" and then nothing happens for a VERY long time, literally minutes, before it finally gives you a plot in the right world. 

Sometimes it can be fast, most of the time its very slow. We do have a huge number of plots in our "plots" world (over 33,000). 

There are no errors on the console.

**Comment by divad** [comment]
That is to say - the [world] parameter works if you're not in a plot world, otherwise it doesn't work and just uses the world you're in.

**Comment by MattBDev** [comment]
if you have over 33,000 plots then it is going to be slow to find a plot no matter what I do. Eventually if you get to 50,000 plots then plot auto will stop working since it really shouldn't be that high.
I fixed the message issue and I am looking into the world argument issue.

**Comment by divad** [comment]
Many thanks

**Comment by jomo** [comment]
couldn't you change the way `/plot auto` works so that it doesn't look for the closest free plot, but instead just looks for the closest free plot **from spawn**? If every *claimed* plot has a record in the DB, you [can use SQL]( http://stackoverflow.com/questions/1389605/sql-find-missing-ids-in-a-table) to find plots that are unclaimed (which might be faster), if all plots from 0;0 to the highest ID are claimed you could just assign the plot that's next to the highest ID, preferably in a (squared) spiral pattern around the spawn.

**Comment by ItsHarry** [comment]
^ An even better solution would be to just keep a list of free plots cached in memory. The cache is updated in a background task when the cache has less than 5 plots left.

I have this exact set up on my own server, and it works absolutely beautifully. Zero lag at all as all the MySQL queries are done in an asynchronous task.

**Comment by divad** [comment]
Thats how 0.16 worked, I believe.

**Comment by ItsHarry** [comment]
No, 0.16 works through a spiral loop, which meant thousands of queries to SQL on the main thread. It caused lag, hence why it was changed in 0.17.
